### PR TITLE
Swap ID types to use NonZeroU64 internally

### PR DESCRIPTION
**Pull request by GnomedDev**
ID 0 is never a valid id, in the case that if you got a 0 id, you would not be able to use it for the API (eg: fetch it as a user).
Therefore, using `NonZeroU64` instead of `u64` allows for rust to use the niche for enum and other optimizations as seen:

serenity-rs/serenity@next
```
std::mem::size_of::<Guild>(); // 896
std::mem::size_of::<Event>(); // 1552
std::mem::size_of::<Channel>(); // 360
std::mem::size_of::<Message>(); // 1000
std::mem::size_of::<VoiceState>(); // 368
std::mem::size_of::<interaction::Interaction>(); // 1544
```
GnomedDev/serenity@nonzerou64
```
std::mem::size_of::<Guild>(); // 848
std::mem::size_of::<Event>(); // 1504
std::mem::size_of::<Channel>(); // 336
std::mem::size_of::<Message>(); // 960
std::mem::size_of::<VoiceState>(); // 352
std::mem::size_of::<interaction::Interaction>(); // 1496
```

I have been running this on [my bot](https://github.com/GnomedDev/Discord-TTS-Bot) in 30k+ servers and haven't had a single panic or deserialization error, so this is safe to run even on large servers or fancy pre-release features.

### CODE REVIEW

**Comment by kangalio** [comment]
Id API changes: **see message below for up-to-date**

- `WhateverId(12345)` => `WhateverId::new(12345)`
- `whatever_id.0` => `whatever_id.get()`
- defining IDs as constants no longer possible (except with unsafe)

File: src/model/id.rs [hunk 1]
```diff
@@ -1,30 +1,48 @@
...
         $(
             impl $name {
+                /// Creates a new Id from a u64
+                ///
+                /// # Panics
+                /// Panics if the id is zero.
+                #[inline]
+                #[must_use]
+                pub fn new(id_as_u64: u64) -> Self {
+                    Self(NonZeroU64::new(id_as_u64).unwrap())
```
Comment (nickelc) [inline src/model/id.rs]: There is no need to encode the type in the variable name.
```suggestion
                pub fn new(id: u64) -> Self {
                    Self(NonZeroU64::new(id).unwrap())
```

File: src/model/id.rs [hunk 2]
```diff
@@ -1,30 +1,48 @@
...
                 pub fn created_at(&self) -> Timestamp {
-                    Timestamp::from_discord_id(self.0)
+                    Timestamp::from_discord_id(self.get())
                 }
 
                 /// Immutably borrow inner Id.
                 #[inline]
                 #[must_use]
-                pub fn as_u64(&self) -> &u64 {
+                pub fn as_u64(&self) -> &NonZeroU64 {
```
Comment (nickelc) [inline src/model/id.rs]: the method name differs from the return type.

File: src/model/id.rs [hunk 3]
```diff
@@ -1,30 +1,48 @@
...
                     &self.0
                 }
 
                 /// Mutably borrow inner Id.
                 #[inline]
                 #[must_use]
-                pub fn as_mut_u64(&mut self) -> &mut u64 {
+                pub fn as_mut_u64(&mut self) -> &mut NonZeroU64 {
```
Comment (nickelc) [inline src/model/id.rs]: the method name differs from the return type.

**Comment by GnomedDev** [comment]
Updated ID API 
- `WhateverId(12345)` => `WhateverId::new(12345)`
- `whatever_id.0` => `whatever_id.get()`
- `whatever_id.as_u64()` => `whatever_id.as_inner()`
- ~~`whatever_id.as_mut_u64`~~
- defining IDs as constants no longer possible (except with unsafe)

File: src/http/client.rs [hunk 4]
```diff
@@ -1758,7 +1756,7 @@ impl Http {
         .await
     }
 
-    /// Follow a News Channel to send messages to a target channel.
+    /// Follow an Announcement Channel to send messages to a target channel.
```
Comment (arqunis) [inline src/http/client.rs]: Unrelated change.

File: src/model/channel/channel_id.rs [hunk 5]
```diff
@@ -379,15 +385,20 @@ impl ChannelId {
...
+                self.get(),
+                message_id.into().get(),
+                &Value::from(map),
+                msg.1,
+            )
             .await
     }
 
-    /// Follows the News Channel
+    /// Follows the announcement channel
```
Comment (arqunis) [inline src/model/channel/channel_id.rs]: Ditto.

**Review by arqunis — CHANGES_REQUESTED** [review]

**Review by arqunis — APPROVED** [review]

**Pull request merged**